File: 1235.maximum-profit-in-job-scheduling/solution.go
func jobScheduling(startTime []int, endTime []int, profit []int) int {
    idx := make([]int, len(startTime))
    for i := range idx {
        idx[i] = i
    }
    sort.Slice(idx, func(i, j int) bool {
        return endTime[idx[i]] < endTime[idx[j]]
    })
    dp := make([]int, len(startTime))
    dp[0] = profit[idx[0]]
    for i := 1; i < len(dp); i++ {
        j := sort.Search(i, func(k int) bool {
            return endTime[idx[k]] > startTime[idx[i]]
        })
        if j > 0 {
            dp[i] = max(dp[i-1], profit[idx[i]] + dp[j-1])
        } else {
            dp[i] = max(dp[i-1], profit[idx[i]])
        }
    }
    return dp[len(dp) - 1]
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: 1235.maximum-profit-in-job-scheduling/solution.kt
class Solution {
    fun jobScheduling(startTime: IntArray, endTime: IntArray, profit: IntArray): Int {
        val n = startTime.size
        val timeOrder = startTime.indices.sortedBy { endTime[it] }.toIntArray()

        fun startTime(job: Int) = startTime[timeOrder[job]]
        fun endTime(job: Int) = endTime[timeOrder[job]]
        fun profit(job: Int) = profit[timeOrder[job]]

        val dp = IntArray(n)
        dp[0] = profit(0)
        for (i in 1 until n) {
            dp[i] = maxOf(dp[i - 1], profit(i))
            for (j in i-1 downTo 0) {
                if (endTime(j) <= startTime(i)) {
                    dp[i] = maxOf(dp[i], dp[j] + profit(i))
                    break
                }
            }
        }
        return dp[n - 1]
    }
}